server/src/main/java/org/opensearch/search/backpressure/trackers/TaskResourceUsageTracker.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.backpressure.trackers;
+
+import org.opensearch.tasks.TaskCancellation;
+import org.opensearch.tasks.Task;
+
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * TaskResourceUsageTracker is used to track completions and cancellations of search related tasks.
+ *
+ * @opensearch.internal
+ */
+public abstract class TaskResourceUsageTracker {
+    /**
+     * Counts the number of cancellations made due to this tracker.
+     */
+    private final AtomicLong cancellations = new AtomicLong();
+
+    public long incrementCancellations() {
+        return cancellations.incrementAndGet();
+    }
+
+    public long getCancellations() {
+        return cancellations.get();
+    }
```

Comment:
Can we delegate this to TaskCancellation instead

Reply:
Here we are tracking the total cancellations done by a task resource usage tracker.

On the other hand, `TaskCancellation` is a wrapper for a single cancellable task along with the list of cancellation reasons and callbacks. The callbacks in it are responsible for calling `incrementCancellations()` on the relevant trackers.

Snippet for reference:
```java
TaskCancellation getTaskCancellation(CancellableTask task) {
    List<TaskCancellation.Reason> reasons = new ArrayList<>();
    List<Runnable> callbacks = new ArrayList<>();

    for (TaskResourceUsageTracker tracker : taskResourceUsageTrackers) {
        Optional<TaskCancellation.Reason> reason = tracker.cancellationReason(task);
        if (reason.isPresent()) {
            reasons.add(reason.get());
            callbacks.add(tracker::incrementCancellations);
        }
    }

    return new TaskCancellation(task, reasons, callbacks);
}
```